File: ui-selenium/src/main/java/org/aeautomation/pages/SignupLoginPage.java
package org.aeautomation.pages;

import org.aeautomation.core.BasePage;
import org.openqa.selenium.By;

/**
 * Page Object representing the Signup / Login portal page (/login).
 */
public class SignupLoginPage extends BasePage {

    // Header assertion locators
    private final By signupHeader = By.xpath("//div[@class='signup-form']/h2");
    private final By loginHeader = By.xpath("//div[@class='login-form']/h2");

    // New User Signup locators
    private final By signupNameInput = By.xpath("//input[@data-qa='signup-name']");
    private final By signupEmailInput = By.xpath("//input[@data-qa='signup-email']");
    private final By signupButton = By.xpath("//button[@data-qa='signup-button']");
    private final By signupErrorMessage = By.xpath("//p[contains(@style,'color: red')]");

    // Existing User Login locators
    private final By loginEmailInput = By.xpath("//input[@data-qa='login-email']");
    private final By loginPasswordInput = By.xpath("//input[@data-qa='login-password']");
    private final By loginButton = By.xpath("//button[@data-qa='login-button']");
    private final By loginErrorMessage = By.xpath("//form[@action='/login']//p");

    public static final String SIGNUP_HEADER_TEXT = "New User Signup!";
    public static final String EXISTING_EMAIL_ERROR_TEXT = "Email Address already exist!";
    public static final String INVALID_CREDENTIALS_ERROR_TEXT = "Your email or password is incorrect!";

    /**
     * Retrieves the text header from the "New User Signup!" section.
     *
     * @return Trimmed text string of the signup header (e.g., "New User Signup!")
     */
    public String getSignupHeaderText() {
        return getText(signupHeader);
    }

    /**
     * Enters the user's name and email into the signup form and submits it.
     *
     * @param name  Full name of the new user
     * @param email Email address for account creation
     * @return New instance of AccountInformationPage for step chaining
     */
    public AccountInformationPage submitSignup(String name, String email) {
        type(signupNameInput, name);
        type(signupEmailInput, email);
        click(signupButton);
        handleGoogleVignetteAd();
        return new AccountInformationPage();
    }

    /**
     * Enters the user's name and email into the signup form when expecting a validation failure.
     *
     * @param name  Full name of the user
     * @param email Already registered email address
     */
    public void submitSignupExpectingFailure(String name, String email) {
        type(signupNameInput, name);
        type(signupEmailInput, email);
        click(signupButton);
        handleGoogleVignetteAd();
    }

    /**
     * Submits the signup form leaving the username field empty.
     *
     * @param email The email address to enter into the signup form
     */
    public void submitSignupWithEmptyUsername(String email) {
        clear(signupNameInput);
        type(signupEmailInput, email);
        click(signupButton);
        handleGoogleVignetteAd();
    }

    /**
     * Retrieves the red validation error message text displayed below the signup form.
     *
     * @return Trimmed error message text (e.g., "Email Address already exist!")
     */
    public String getSignupErrorMessage() {
        return getText(signupErrorMessage);
    }

    /**
     * Retrieves the HTML5 native validation message from the signup email input field.
     *
     * @return The browser's native validation string
     */
    public String getSignupEmailValidationMessage() {
        return getValidationMessage(signupEmailInput);
    }

    /**
     * Retrieves the HTML5 native validation message from the signup username input field.
     *
     * @return The browser's native validation string (e.g., "Please fill out this field.")
     */
    public String getSignupUsernameValidationMessage() {
        return getValidationMessage(signupNameInput);
    }

    /**
     * Logs in with the provided email and password.
     *
     * @param email    User email address
     * @param password User password
     * @return HomePage instance after submitting credentials
     */
    public HomePage login(String email, String password) {
        type(loginEmailInput, email);
        type(loginPasswordInput, password);
        click(loginButton);
        handleGoogleVignetteAd();
        return new HomePage();
    }

    /**
     * Enters the email and password into the login form and submits it when expecting a authentication failure.
     *
     * @param email    User email address
     * @param password User password
     */
    public void loginExpectingFailure(String email, String password) {
        type(loginEmailInput, email);
        type(loginPasswordInput, password);
        click(loginButton);
    }

    /**
     * Checks if the login error message is displayed.
     *
     * @return true if error message is visible, false otherwise
     */
    public boolean isLoginErrorMessageDisplayed() {
        return isDisplayed(loginErrorMessage);
    }

    /**
     * Retrieves the text of the login error message.
     *
     * @return String containing the error text
     */
    public String getLoginErrorMessageText() {
        return getText(loginErrorMessage);
    }

    /**
     * Retrieves the HTML5 native validation message (tooltip) for the login email field.
     *
     * @return HTML5 validation message string (e.g., "Please fill out this field.")
     */
    public String getLoginEmailValidationMessage() {
        return getValidationMessage(loginEmailInput); // Calls your BasePage method
    }

    /**
     * Retrieves the HTML5 native validation message (tooltip) for the login password field.
     *
     * @return HTML5 validation message string (e.g., "Please fill out this field.")
     */
    public String getLoginPasswordValidationMessage() {
        return getValidationMessage(loginPasswordInput);
    }
}

File: ui-playwright/pages/SignupLoginPage.ts
import { Page, Locator } from '@playwright/test';
import { BasePage } from './BasePage';
import { AccountInformationPage } from './AccountInformationPage';

/**
 * Page Object representing the Signup / Login page (`/login`).
 */
export class SignupLoginPage extends BasePage {

    /** Expected header text for the new user registration form. */
    public static readonly SIGNUP_HEADER_TEXT = 'New User Signup!';

    /** Expected header text for the user login form. */
    public static readonly LOGIN_HEADER_TEXT = 'Login to your account';

    /** Error text displayed when attempting to register an already registered email. */
    public static readonly EXISTING_EMAIL_ERROR_TEXT = 'Email Address already exist!';

    /** Error text displayed when attempting to log in with invalid credentials. */
    public static readonly INVALID_CREDENTIALS_ERROR_TEXT = 'Your email or password is incorrect!';

    /** Heading locator for the signup section (`getByRole`). */
    readonly newUserSignupHeader: Locator;

    /** Text input locator for the signup name field (`getByTestId`). */
    readonly signupNameInput: Locator;

    /** Text input locator for the signup email field (`getByTestId`). */
    readonly signupEmailInput: Locator;

    /** Button locator to submit the signup form (`getByTestId`). */
    readonly signupButton: Locator;

    /** Element locator for signup error message text (`getByText`). */
    readonly signupErrorMessageText: Locator;

    /** Heading locator for the login section (`getByRole`). */
    readonly loginHeader: Locator;

    /** Text input locator for the login email field (`getByTestId`). */
    readonly loginEmailInput: Locator;

    /** Text input locator for the login password field (`getByTestId`). */
    readonly loginPasswordInput: Locator;

    /** Button locator to submit the login form (`getByTestId`). */
    readonly loginButton: Locator;

    /** Element locator for login error message text (`getByText`). */
    readonly loginErrorMessageText: Locator;

    constructor(page: Page) {
        super(page);

        // Section Headings (Semantic Role)
        this.newUserSignupHeader = page.getByRole('heading', { name: SignupLoginPage.SIGNUP_HEADER_TEXT });
        this.loginHeader = page.getByRole('heading', { name: SignupLoginPage.LOGIN_HEADER_TEXT });

        // Signup Form Controls (data-qa test IDs for fast execution)
        this.signupNameInput = page.getByTestId('signup-name');
        this.signupEmailInput = page.getByTestId('signup-email');
        this.signupButton = page.getByTestId('signup-button');

        // Login Form Controls (data-qa test IDs for fast execution)
        this.loginEmailInput = page.getByTestId('login-email');
        this.loginPasswordInput = page.getByTestId('login-password');
        this.loginButton = page.getByTestId('login-button');

        // Error Messages (Visible Copy)
        this.loginErrorMessageText = page.getByText(SignupLoginPage.INVALID_CREDENTIALS_ERROR_TEXT);
        this.signupErrorMessageText = page.getByText(SignupLoginPage.EXISTING_EMAIL_ERROR_TEXT);
    }

    /**
     * Fills the new user registration input fields with a name and email address.
     * 
     * @param name - Full name or username for registration
     * @param email - Email address for the new account
     */
    async enterSignupDetails(name: string, email: string): Promise<void> {
        await this.type(this.signupNameInput, name);
        await this.type(this.signupEmailInput, email);
    }

    /**
     * Clicks the 'Signup' button to submit initial registration credentials.
     * 
     * @returns Promise resolving to a new instance of AccountInformationPage
     */
    async clickSignup(): Promise<AccountInformationPage> {
        await this.click(this.signupButton);
        return new AccountInformationPage(this.page);
    }

    /**
     * Fills credentials and submits the login form.
     * 
     * @param email - User account email address
     * @param password - User account password
     */
    async login(email: string, password: string): Promise<void> {
        await this.type(this.loginEmailInput, email);
        await this.type(this.loginPasswordInput, password);
        await this.click(this.loginButton);
    }

    /**
     * Retrieves the native HTML5 browser validation message from the signup email input field.
     * 
     * @returns Promise resolving to the HTML5 validation message string
     */
    async getSignupEmailValidationMessage(): Promise<string> {
        return this.getValidationMessage(this.signupEmailInput);
    }

    /**
     * Retrieves the native HTML5 browser validation message from the signup name input field.
     * 
     * @returns Promise resolving to the HTML5 validation message string
     */
    async getSignupUsernameValidationMessage(): Promise<string> {
        return this.getValidationMessage(this.signupNameInput);
    }

    /**
     * Retrieves the native HTML5 browser validation message from the login email input field.
     * 
     * @returns Promise resolving to the HTML5 validation message string
     */
    async getLoginEmailValidationMessage(): Promise<string> {
        return this.getValidationMessage(this.loginEmailInput);
    }

    /**
     * Retrieves the native HTML5 browser validation message from the login password input field.
     * 
     * @returns Promise resolving to the HTML5 validation message string
     */
    async getLoginPasswordValidationMessage(): Promise<string> {
        return this.getValidationMessage(this.loginPasswordInput);
    }
}